rclpy/src/rclpy/context.cpp

```diff
@@ -150,17 +150,18 @@ Context::ok()
 void
 Context::shutdown()
 {
-  {
-    std::lock_guard<std::mutex> guard{g_contexts_mutex};
-    auto iter = std::find(g_contexts.begin(), g_contexts.end(), rcl_context_.get());
-    if (iter != g_contexts.end()) {
-      g_contexts.erase(iter);
-    }
+  std::lock_guard<std::mutex> guard{g_contexts_mutex};
+  if (already_shutdown_) {
+    throw CONTEXT_ALREADY_SHUTDOWN("Context already shutdown.");
   }
-
-  rcl_ret_t ret = rcl_shutdown(rcl_context_.get());
-  if (RCL_RET_OK != ret) {
-    throw RCLError("failed to shutdown");
+  auto iter = std::find(g_contexts.begin(), g_contexts.end(), rcl_context_.get());
```

Comment:
so the `g_contexts` is the collection of valid context, and this means if it cannot find the context in the `g_contexts` during this shutdown call, that is the  invalid operation? so we can generate the exception without introducing `already_shutdown_`?

Reply:
> so the g_contexts is the collection of valid context, and this means if it cannot find the context in the g_contexts during this shutdown call, that is the invalid operation?

Yes.  
`rc_context_` is added to g_contexts in constructor of Context.  Before calling rcl_shutdown(), `rc_context_` is removed from g_contexts.

> so we can generate the exception without introducing already_shutdown_?

`already_shutdown_` is only used for multiple calls to Context::shutdown() to throw the exception.   

About issue reported by https://github.com/ros2/rclpy/issues/1352, it describes a possible scenario. There are two threads to call rcl_shutdown for the same context. One call is from rclpy::shutdown() and the other from Context::shutdown(). In this situation, we do not want to get an exception in Context::shutdown().

Reply:
that is exactly why we have `std::lock_guard<std::mutex> guard{g_contexts_mutex};` lock escalated to the this function scope? so that when [rclpy::shutdown_contexts](https://github.com/ros2/rclpy/blob/8a2e2d3de5a96eaa50b9c9336d97396f8489bda4/rclpy/src/rclpy/context.cpp#L49) is under process, we will not find the iterator with this context, so generate the exception. i really do not understand to have `already_shutdown_` flag internally...

Reply:
@fujitatomoya 

> i really do not understand to have already_shutdown_ flag internally...

Let me first summarize the two scenarios we currently need to address:

1. The scenario mentioned in #1352. There are two threads to call rcl_shutdown for the same context. One call is from rclpy::shutdown() and the other from Context::shutdown(). For this scenario, we don't want to get exception.   
2. For one context instance, Context::shutdown() is being called multiple times. At this scenario, we want to throw an exception. (There is an existed test to check if an exception is thrown.)  The fix for 1 would prevent 2 from throwing an exception. That's why already_shutdown_ ``` was added to indicate that Context::shutdown() has already been called once on the context instance. Do you have any better suggestions?

Reply:
@fujitatomoya Friendly ping.

Reply:
thanks for the explanation, i understand what you mean. if we take `already_shutdown_`, then we need to reinitialize `already_shutdown_` during context initialization? unless, `already_shutdown_` will never be False?

IMO, `Context::shutdown()` should generate the exception only when `rcl_shutdown` fails with valid context. this actually aligns with `rclcpp` behavior. of course, if we change like this, we need to adjust the unit tests accordingly.

maybe we can keep this open for more feedback.

Reply:
> thanks for the explanation, i understand what you mean. if we take already_shutdown_, then we need to reinitialize already_shutdown_ during context initialization? unless, already_shutdown_ will never be False?

Yes. 

> IMO, Context::shutdown() should generate the exception only when rcl_shutdown fails with valid context. this actually aligns with rclcpp behavior. of course, if we change like this, we need to adjust the unit tests accordingly.

Yes. The addition of already_shutdown_ is just to pass the existing tests (Ensure consistency with previous behavior.)

> maybe we can keep this open for more feedback.

Yes.